qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -266,63 +251,198 @@ def run(self, dag):
+            _apply_sabre_result(
+                mapped_dag,
+                dag,
+                self._qubit_indices,
+                canonical_register,
+                original_layout,
+                sabre_result,
+                circuit_to_dag_dict,
+            )
+            return mapped_dag
+        return dag
+
+
+def _build_sabre_dag(dag, num_qubits, num_clbits, qubit_indices, clbit_indices):
+    from qiskit.converters import circuit_to_dag
+
+    # Maps id(block): circuit_to_dag(block) for all descendant blocks
+    circuit_to_dag_dict = {}
+
+    def recurse(block, block_qubit_indices, block_clbit_indices):
+        block_dag = circuit_to_dag(block)
+        circuit_to_dag_dict[id(block)] = block_dag
+        return process_dag(block_dag, block_qubit_indices, block_clbit_indices)
+
+    def process_dag(block_dag, block_qubit_indices, block_clbit_indices):
+        dag_list = []
+        node_blocks = {}
+        for node in block_dag.topological_op_nodes():
+            cargs = {block_clbit_indices[x] for x in node.cargs}
+            if node.op.condition is not None:
+                for clbit in block_dag._bits_in_condition(node.op.condition):
+                    cargs.add(block_clbit_indices[clbit])
+            if isinstance(node.op, ControlFlowOp):
+                node_blocks[node._node_id] = [
+                    recurse(
+                        block,
+                        {
+                            inner: block_qubit_indices[outer]
+                            for inner, outer in zip(block.qubits, node.qargs)
+                        },
+                        block_clbit_indices,
+                    )
+                    for block in node.op.blocks
+                ]
+            dag_list.append(
+                (
+                    node._node_id,
+                    [block_qubit_indices[x] for x in node.qargs],
+                    cargs,
                 )
-                apply_gate(
-                    mapped_dag,
-                    node,
-                    original_layout,
+            )
+
+        # Note: num_qubits and num_clbits are used to preallocate
+        # storage within SabreDAG. Because indices are in terms of the
+        # root DAG, we always pass the same values even though blocks
+        # will often have fewer bits than the outer circuit.
+        return SabreDAG(num_qubits, num_clbits, dag_list, node_blocks)
+
+    return process_dag(dag, qubit_indices, clbit_indices), circuit_to_dag_dict
+
+
+def _apply_sabre_result(
+    mapped_dag,
+    root_dag,
+    qubit_indices,
+    canonical_register,
+    initial_layout,
+    sabre_result,
+    circuit_to_dag_dict,
+    component_map=None,
+):
+    bit_to_qreg_idx = {bit: idx for idx, bit in enumerate(canonical_register)}
+
+    def empty_dag(node):
+        out = DAGCircuit()
+        for qreg in mapped_dag.qregs.values():
+            out.add_qreg(qreg)
+        out.add_clbits(node.cargs)
+        return out
```

Comment:
I'm concerned that some inner blocks may have things like global phases or classical registers (for example, if a control-flow block itself contains a control-flow block with a `ClassicalRegister` condition), which this conversion will miss.

Reply:
Ah good to know. I was under the impression that inner blocks shouldn't have their own global properties or registers. I suppose we can add the current block as a second arg and copy these things from that.

Reply:
Ah alright, so the biggest issue here is that I'm not adding the `cregs` from the `mapped_dag`, correct? I can also take copy in any global phases from the current block as well.

Reply:
Yeah, I think `cregs` is the big one.  We should probably include a test that a nested block that contains a classical register all works correctly as well - if you need to compare, you can use the control-flow builder interface, then make sure that the inner blocks use register conditions rather than bit conditions, and it should create control-flow ops whose blocks involve registers as well.

Reply:
Fixed in 386197d.